### PR TITLE
Add toInterface

### LINKED ISSUE #178: fluent: provide an API to convert an ipld.Node to a generic Go value
In other words, the reverse of `fluent.Reflect`. `Int` gives an `int64`, `Map` gives a `map[string]interface{}`, `List` gives a `[]interface{}`, and so on.

This is somewhat similar to how code-generated scalar nodes have methods like `Int() int64` as well as the usual `AsInt() (int64, error)`, but applied more generally for any node implementation and kind.

The main use case is when one has small IPLD nodes, such as any scalar kind or a small list/map, and one wants to obtain the plain Go value for easy use. Especially for debugging or testing, too.

This could potentially allocate a lot for large maps or lists, but we assume noone would do that.

Unfortunately this does mean that calling this API on a `List_String` would return a `[]interface{}` instead of `[]string`, which is unfortunate and not terribly helpful. But I don't think we can reasonably do better right now. Perhaps we can revisit that with generics.

cc @warpfork @willscott for the recent discussion
cc @gammazero since your recent question made me reconsider proposing this

### PR DESCRIPTION
**Pull request by MarcoPolo**
Fixes #178 

Does this seem like the general idea you were thinking of @mvdan? If so I'll add more documentation and maybe a couple more test cases.

### CODE REVIEW

File: fluent/toInterfaceValue.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
+package fluent
+
+import (
+	"github.com/ipld/go-ipld-prime/datamodel"
+)
+
+func ToInterfaceValue(node datamodel.Node) (interface{}, error) {
+	switch k := node.Kind(); k {
+	case datamodel.Kind_Invalid:
+		panic("invalid node")
```
Comment (mvdan) [inline fluent/toInterfaceValue.go:10]: Panicking here seems a bit harsh :) An error woudl be better. Otherwise one would have to wrap most calls to this func with a recover, because it's hard to tell in advance if some input data or edge case program behavior could lead to an invalid kind.

File: fluent/toInterfaceValue.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
+package fluent
+
+import (
+	"github.com/ipld/go-ipld-prime/datamodel"
+)
+
+func ToInterfaceValue(node datamodel.Node) (interface{}, error) {
```
Comment (mvdan) [inline fluent/toInterfaceValue.go:7]: After having slept on my suggestion of ToInterfaceValue, I wonder if just ToInterface would be enough. All functions that return `(T, err)` already return a value.

File: fluent/toInterfaceValue.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+	case datamodel.Kind_Float:
+		return node.AsFloat()
+	case datamodel.Kind_String:
+		return node.AsString()
+	case datamodel.Kind_Bytes:
+		return node.AsBytes()
+	case datamodel.Kind_Link:
+		return node.AsLink()
+	case datamodel.Kind_Map:
+		outMap := make(map[string]interface{})
```
Comment (mvdan) [inline fluent/toInterfaceValue.go:26]: nit: I think you can also use a size hint here

File: fluent/toInterfaceValue.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+		for mi := node.MapIterator(); !mi.Done(); {
+			k, v, err := mi.Next()
+			if err != nil {
+				return nil, err
+			}
+			kVal, err := k.AsString()
```
Comment (mvdan) [inline fluent/toInterfaceValue.go:32]: perhaps leave a comment explaining that, for now, we assume all keys are strings.

it's an okay assumption to make right now, but we should also make it clear.

File: fluent/toInterfaceValue.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+			if err != nil {
+				return nil, err
+			}
+			vVal, err := ToInterfaceValue(v)
+			if err != nil {
+				return nil, err
+			}
+			outMap[kVal] = vVal
+
+			if mi.Done() {
```
Comment (mvdan) [inline fluent/toInterfaceValue.go:42]: isn't this check duplicated?

File: fluent/toInterfaceValue.go [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+			if err != nil {
+				return nil, err
+			}
+			vVal, err := ToInterfaceValue(v)
+			if err != nil {
+				return nil, err
+			}
+			outList = append(outList, vVal)
+
+			if li.Done() {
```
Comment (mvdan) [inline fluent/toInterfaceValue.go:60]: ditto

File: fluent/toInterfaceValue.go [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+				break
+			}
+		}
+		return outList, nil
+	}
+	panic("unhandled case in switch")
```
Comment (mvdan) [inline fluent/toInterfaceValue.go:66]: nit: move this under `default:`, that way you ensure you don't end up in this panic in any other of the `case`s

File: fluent/toInterfaceValue_test.go [hunk 8]
```diff
@@ -0,0 +1,55 @@
+package fluent_test
+
+import (
+	"testing"
+
+	"github.com/ipld/go-ipld-prime/fluent"
+	"github.com/ipld/go-ipld-prime/node/basicnode"
+)
+
+func TestListValue(t *testing.T) {
```
Comment (mvdan) [inline fluent/toInterfaceValue_test.go:10]: I'd encourage you to use table-driven tests, to avoid duplication. This way you can also have code coverage for all edge cases by just adding a few more entries to the table.

For instance: https://github.com/ipld/go-ipld-prime/blob/892d2f92ba8bdecef98d8863d310dbbd002b726c/node/bindnode/infer_test.go#L41

File: fluent/toInterfaceValue_test.go [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+	outM := out.(map[string]interface{})
+
+	if len(a) != len(outM) {
+		t.Errorf("Mismatch in size")
+	}
+
+	for k, v := range outM {
+		if v != a[k] {
+			t.Errorf("expected %v, got %v at key %v", a[k], v, k)
+		}
```
Comment (mvdan) [inline fluent/toInterfaceValue_test.go:52]: You'd save on keystrokes, and have better errors, if you use the qt testing library :) Like @masih did in PRs like https://github.com/ipld/go-ipld-prime/pull/291.

It's not a must to refactor all the code right now, but we might as well use it in new code.

**Review by mvdan — COMMENTED** [review]
Looks good! The comments are mostly minor.

File: fluent/toInterfaceValue.go (hunk 4 above)
Comment (MarcoPolo) [reply fluent/toInterfaceValue.go:32]: My understanding is that IPLD maps always have string keys. It says this in the docs:
https://ipld.io/docs/intro/primer/#finer-points
> Map keys are strings.

Comment (mvdan) [reply fluent/toInterfaceValue.go:37]: Ah, you're right, the docs even say that!

File: fluent/toInterfaceValue_test.go [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+		})
+	}
+}
+
+func TestLink(t *testing.T) {
+	c := qt.New(t)
+	var someCid, err = cid.Parse("bafybeihrqe2hmfauph5yfbd6ucv7njqpiy4tvbewlvhzjl4bhnyiu6h7pm")
+	c.Assert(err, qt.IsNil)
+	var link = cidlink.Link{Cid: someCid}
+	var node = basicnode.NewLink(link)
```
Comment (mvdan) [inline fluent/toInterfaceValue_test.go:46]: for consistency, use `foo := bar` rather than `var foo = bar`; the latter is generally only used when you want to explicitly declare the type - `var foo Type = bar` - or when you're declaring globals, where you can't just use `:=` statements :)

File: fluent/toInterfaceValue.go [hunk 11]
```diff
@@ -0,0 +1,65 @@
+package fluent
+
+import (
+	"errors"
+
+	"github.com/ipld/go-ipld-prime/datamodel"
+)
+
+var ErrInvalidKind = errors.New("invalid kind")
+var ErrUnknownKind = errors.New("unknown kind")
```
Comment (mvdan) [inline fluent/toInterfaceValue.go:10]: nit: I would avoid exposing these errors in the API for now, because once they are exposed, they are part of the API we must support. These errors should be very rare and almost always due to a bug that needs to be fixed in someone's code, so it also doesn't seem like errors that the end user should be checking against.

File: fluent/toInterfaceValue.go [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+
+func ToInterface(node datamodel.Node) (interface{}, error) {
```
Comment (mvdan) [inline fluent/toInterfaceValue.go:12]: Needs docs :) for instance:

```
// ToInterface converts an IPLD node to its simplest Go value equivalent.
//
// Booleans, integers, floats, strings, bytes, and links are returned as themselves,
// as per the node's AsT method. Note that nulls are returned as untyped nils.
//
// Lists and maps are returned as []interface{} and map[string]interface{}, respectively.
```

File: fluent/toInterfaceValue_test.go [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+
+	qt "github.com/frankban/quicktest"
+	"github.com/ipfs/go-cid"
+	"github.com/ipld/go-ipld-prime/fluent"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+	"github.com/ipld/go-ipld-prime/node/basicnode"
+)
+
+var roundTripTestCases = []struct {
+	desc  string
```
Comment (mvdan) [inline fluent/toInterfaceValue_test.go:15]: nit: usually you'd name this `name`, since it's the sub-test name, not really its description

File: fluent/toInterfaceValue_test.go [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+	value interface{}
+}{
+	{desc: "Number", value: 100},
+	{desc: "String", value: "hi"},
+	{desc: "Bool", value: "hi"},
```
Comment (mvdan) [inline fluent/toInterfaceValue_test.go:20]: did you mean to make this a boolean? :)

if you use `go test -coverprofile=cover.out && go tool cover -html=cover.out`, you'll be able to switch to your file and see what lines aren't being tested at all.

File: fluent/toInterfaceValue_test.go [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+	for _, testCase := range roundTripTestCases {
+		t.Run(testCase.desc, func(t *testing.T) {
+			c := qt.New(t)
+			n, err := fluent.Reflect(basicnode.Prototype.Any, testCase.value)
+			c.Assert(err, qt.IsNil)
+			out, err := fluent.ToInterface(n)
+			c.Assert(err, qt.IsNil)
+			outJson, err := json.Marshal(out)
+			qt.Check(t, err, qt.IsNil)
+			qt.Check(t, outJson, qt.JSONEquals, testCase.value)
```
Comment (mvdan) [inline fluent/toInterfaceValue_test.go:36]: nit: you should be able to use `qt.DeepEquals`, I think. the only other change you'd need is for the Array test case to use `[]interface{}` as its type.

**Review by mvdan — APPROVED** [review]
Looks good! Just some minor nits, we can merge when they're addressed.

File: fluent/toInterfaceValue_test.go (hunk 15 above)
Comment (MarcoPolo) [reply fluent/toInterfaceValue_test.go:36]: 👍 I was wondering if that would work after I made this change. `qt` is pretty neat.

File: fluent/toInterfaceValue_test.go (hunk 14 above)
Comment (MarcoPolo) [reply fluent/toInterfaceValue_test.go:20]: nice! I didn't know that

**Comment by MarcoPolo** [comment]
Thanks for the great comments @mvdan

**Review by mvdan — APPROVED** [review]
Nice work :)
